massa-execution-worker/src/speculative_roll_state.rs

```diff
@@ -476,4 +507,55 @@ impl SpeculativeRollState {
 
         credits
     }
+
+    /// Remove a given amount from deferred credits and return the removed amount
+    pub fn remove_deferred_credits(&mut self, slot: &Slot, addr: &Address, amount: &Amount) -> Option<Amount> {
+
+        let mut amount_to_rm = *amount;
+        let mut amount_removed = Amount::zero();
+
+        // First try to rm in added_changes
+        let res = self.added_changes.deferred_credits.sub_amount(slot, addr, &amount_to_rm);
+        if let Some(a) = res {
+            amount_removed = amount_removed.checked_add(a)?;
+            amount_to_rm = amount_to_rm.checked_sub(amount_removed)?;
+        }
+
+        // If not found, try to rm from active history
+        let mut hist = self.active_history.write();
+        while amount_removed < *amount {
+            match hist.0.iter_mut().rev().next() {
+                None => break,
+                Some(hist_item) => {
+                    let res = hist_item
+                        .state_changes
+                        .pos_changes
+                        .deferred_credits
+                        .sub_amount(slot, addr, &amount_to_rm);
+
+                    if let Some(a) = res {
+                        amount_removed = amount_removed.checked_add(a)?;
+                        amount_to_rm = amount_to_rm.checked_sub(amount_removed)?;
+                    }
+                }
+            }
+        }
+        drop(hist);
+
+        // At the end, try in final state
+        if amount_removed < *amount {
+            let res = self
+                .final_state
+                .write()
+                .pos_state
+                .rm_deferred_credits(slot, addr, amount);
+
+            if let Some(a) = res {
+                amount_removed = amount_removed.checked_add(a)?;
+                // No need to update amount_to_tm here
+            }
+        }
```

Comment:
In my unit tests, the deferred credits are in added_changes but I dunno if it's relevant from the 'real' cases. What do you think @damip ?

Reply:
> We still slash in the same order (earliest-to-be-credited first)

But we don't here do we? We slash in changes > active > final. So we're taking away the last that would be reimbursed since we go through the more recent credits and it's always 3 cycles for reimbursement.

> And we can't only look into the final state of deferred credits when creating a candidate slashing, otherwise we might end up attempting to slash multiple times something that was already slashed in the active history but not in the final state yet.

Not sure I understand this, if we keep the current behaviour meaning we go through the credits and stop iterating once we reach the amount that was to be slashed that won't happen no?

> We can't slash the final state from a candidate slot or we will run into consensus problems.

I was speaking of setting the slashing as execution output apply it to final_state if the times comes. But as a matter of fact I just noticed that we currently write on final state in the `speculative_roll_state` which is never supposed happen because the eventual changes can't be undone. final_state in the "speculative" versions is only meant to be read, I think we should consider using a read only lock to avoid this in the future.